llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp

```diff
@@ -1026,17 +1027,15 @@ void SystemZAsmPrinter::emitADASection() {
       EmittedBytes += PointerSize;
       break;
     case SystemZII::MO_ADA_INDIRECT_FUNC_DESC: {
-      MCSymbol *Alias = OutContext.createTempSymbol(
-          Twine(Sym->getName()).concat("@indirect"));
-      OutStreamer->emitAssignment(Alias,
-                                  MCSymbolRefExpr::create(Sym, OutContext));
-      OutStreamer->emitSymbolAttribute(Alias, MCSA_IndirectSymbol);
+      MCSymbolGOFF *IndirectSym =
+          static_cast<MCSymbolGOFF *>(OutContext.getOrCreateSymbol(
+              Twine(Sym->getName()).concat("@indirect")));
```

Comment:
This now just creates a completely new symbol, that is totally unrelated to the original symbol `Name`.   How is that supposed to work?  Where will this be associated back to the original symbol and its address?

Reply:
One of the issues is that using the alias construct isn't entirely accurate. If Symbol A aliases Symbol B, then uses of A can (and are) be replaced by uses of B.

However, this isn't really the case here. The `MO_ADA_INDIRECT_FUNC_DESC` here means that we have a symbol that has a value containing the location of the function descriptor of the symbol it points to. 

![onno_art_is_hard](https://github.com/llvm/llvm-project/assets/22278196/463325c8-738b-4837-b648-a5fec8945998)

What we want to do here is to create a new symbol that points to a location in the ADA. That location contains the offset in the ADA to the location where the original symbol's ADA entry resides.

The GOFF writer creates an external reference (ErWx) symbol to represent this newly-created symbol in our symbol table (Specifically, the "external symbol dictionary"). This new symbol will have the same name as the original symbol, and that name is used by the binder to associate it back to the original symbol's address).